### PR TITLE
feat: cross chain swap STX

### PR DESCRIPTION
**Pull request by infiniteflower**
## **Description**

This PR improves support for Smart Transactions (STX). It will not show the STX status page if the tx is a Bridge tx. Instead users will be immediately sent to the Activity list. If the user opens the Bridge tx detaills, they should be able to see the proper info.

The main change is that in `BridgeStatusController` the `txHistory` is now keyed by `txMeta.id` rather than `txMeta.hash`. This is because for an STX we might not have the hash immediately, but we still want to look things up.

We also add a way to call `addTransaction` from the UI side. Previously we used `addTransactionAndWaitForPublish`. However for STX, this will cause the app to wait until a `txHash` is available before returning a `txMeta`, which means we can't even use the `txMeta.id` to key a `BridgeHistoryItem`. If we use `addTransaction` this will allow us to receive a `txMeta` object back immediately and use `txMeta.id`.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/28460?quickstart=1)

## **Related issues**

Depends on https://github.com/MetaMask/core/pull/4918

## **Manual testing steps**

1. Add `BRIDGE_USE_DEV_APIS=1` to `.metamaskrc`
1. Go to Settings > Advanced, make sure Smart Transactions is on
2. Switch network to Ethereum
3. Go to Bridge
4. Selection source network, source token, dest network, dest token
5. Request a quote
6. Execute a bridge
7. Go to Activity
8. Click on the Bridge tx

### **After**

https://github.com/user-attachments/assets/364b93ec-3b54-4ce6-acd5-5001236c128c

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/controllers/bridge-status/bridge-status-controller.ts [hunk 1]
```diff
@@ -264,14 +276,61 @@ export default class BridgeStatusController extends StaticIntervalPollingControl
...
+    const nextSrcTxHash =
+      nextState.bridgeStatusState.txHistory[bridgeTxMetaId].status.srcChain
+        .txHash;
+
+    if (nextSrcTxHash) {
+      return nextSrcTxHash;
+    }
+
+    throw new Error('srcTxHash missing');
+  };
```
Comment (micaelae) [inline app/scripts/controllers/bridge-status/bridge-status-controller.ts:324]: Do we really need to do a state update in here? Are there downsides to just returning `txMeta?.hash` and letting `#fetchBridgeTxStatus` handle it?

If there's no hash fetchBridgeTxStatus can return early

Comment (micaelae) [reply app/scripts/controllers/bridge-status/bridge-status-controller.ts:324]: Is it typical for a smart tx to not have a hash right after submission? If it's expected, I also don't think we should throw an error if it's not found. This helps keep our logs/errors actionable

File: ui/pages/bridge/transaction-details/bridge-step-list.tsx [hunk 2]
```diff
@@ -85,7 +86,7 @@ export default function BridgeStepList({
             isEdgeComplete={isEdgeComplete}
           >
             <BridgeStepDescription
-              step={step}
+              step={step as Step}
```
Comment (micaelae) [inline ui/pages/bridge/transaction-details/bridge-step-list.tsx:89]:
```suggestion
              step={step}
```


Do you need the type assertion here?

File: ui/pages/bridge/transaction-details/transaction-details.tsx [hunk 3]
```diff
@@ -60,6 +63,34 @@ const getBlockExplorerUrl = (
...
+const getBridgeAmount = ({
+  bridgeHistoryItem,
+  srcChainTxMeta,
+}: {
+  bridgeHistoryItem?: BridgeHistoryItem;
+  srcChainTxMeta?: TransactionMeta;
+}) => {
+  if (bridgeHistoryItem) {
+    return `${calcTokenAmount(
+      bridgeHistoryItem.quote.srcTokenAmount,
```
Comment (micaelae) [inline ui/pages/bridge/transaction-details/transaction-details.tsx:75]: This won't include the MM fee (part of the actual transferred amount)

If sentAmount and other [calculated quote metadata](https://github.com/MetaMask/metamask-extension/blob/develop/ui/ducks/bridge/selectors.ts#L233) fields are saved to the txHistory `pricingData` field, we can use `sentAmount` here instead. We should audit other usages of `srcTokenAmount` as well and use `sentAmount` too

File: ui/pages/bridge/transaction-details/bridge-step-list.tsx (hunk 2 above)
Comment (infiniteflower) [reply ui/pages/bridge/transaction-details/bridge-step-list.tsx]: Removed in [d49cbcf](https://github.com/MetaMask/metamask-extension/pull/28460/commits/d49cbcf83b1950a4de52c0cfb7388fb4fdbd8a8c)

File: app/scripts/controllers/bridge-status/bridge-status-controller.ts (hunk 1 above)
Comment (infiniteflower) [reply app/scripts/controllers/bridge-status/bridge-status-controller.ts:330]: Done in [c72ac4e](https://github.com/MetaMask/metamask-extension/pull/28460/commits/c72ac4efe3d3828ff8146a1ff385d110e9ff6800)

File: ui/pages/bridge/transaction-details/transaction-details.tsx (hunk 3 above)
Comment (infiniteflower) [reply ui/pages/bridge/transaction-details/transaction-details.tsx:75]: > Should probably include at least sentAmount in the txMeta as well

Confirmations team's recommendation was to store very bridge specific data in its own controller, so I think that putting it into `txHistory.pricingData` should be sufficient.

Comment (infiniteflower) [reply ui/pages/bridge/transaction-details/transaction-details.tsx]: Done in [b52bf65](https://github.com/MetaMask/metamask-extension/pull/28460/commits/b52bf650be57cd4224149f178fe7cb6f15a2a273) and https://github.com/MetaMask/metamask-extension/pull/28460/commits/110c473a43a69ce99b4bdf770a9a615de52c010b

**Review by micaelae — DISMISSED** [review]

File: app/scripts/lib/transaction/smart-transactions-mocks.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+export const mockSignAndSubmitTransactions: () => Promise<{
+  uuid: string;
+  txHash?: string;
+}> = () => {
+  return new Promise((resolve) => {
+    setTimeout(() => {
+      resolve({
+        uuid: 'uuid123456789',
+      });
+    }, 100);
```
Comment (dan437) [inline app/scripts/lib/transaction/smart-transactions-mocks.ts:33]: These functions are not supposed to be used in unit / integration tests, right?

File: app/scripts/lib/transaction/smart-transactions.ts [hunk 5]
```diff
@@ -169,12 +174,15 @@ class SmartTransactionHook {
...
       const extensionReturnTxHashAsap =
         this.#featureFlags?.smartTransactions?.extensionReturnTxHashAsap;
-      this.#addApprovalRequest({
-        uuid,
-      });
-      this.#addListenerToUpdateStatusPage({
-        uuid,
-      });
+
+      if (shouldShowStatusPage) {
```
Comment (dan437) [inline app/scripts/lib/transaction/smart-transactions.ts:178]: Shouldn't this condition be also in the `#onApproveOrReject` function for
```
this.#controllerMessenger.call('ApprovalController:endFlow', {
  id: this.#approvalFlowId,
});
```

File: app/scripts/lib/transaction/smart-transactions-mocks.ts (hunk 4 above)
Comment (infiniteflower) [reply app/scripts/lib/transaction/smart-transactions-mocks.ts:33]: No it was just for testing out the flow without paying gas. I left them there for convenience sake, but we can take them out if you would like.

File: app/scripts/lib/transaction/smart-transactions.ts (hunk 5 above)
Comment (infiniteflower) [reply app/scripts/lib/transaction/smart-transactions.ts]: Done in [03c74fd](https://github.com/MetaMask/metamask-extension/pull/28460/commits/03c74fd891bbda3ff3e0ad92e0f2aa00b644c19a)

File: app/scripts/lib/transaction/smart-transactions.ts [hunk 6]
```diff
@@ -123,6 +125,8 @@ class SmartTransactionHook {
     this.#isDapp = transactionMeta.origin !== ORIGIN_METAMASK;
     this.#chainId = transactionMeta.chainId;
     this.#txParams = transactionMeta.txParams;
+    this.#shouldShowStatusPage =
+      transactionMeta.type !== TransactionType.bridge;
```
Comment (micaelae) [inline app/scripts/lib/transaction/smart-transactions.ts:129]: Is this needed for `TransactionType.bridgeApproval` as well?

Comment (infiniteflower) [reply app/scripts/lib/transaction/smart-transactions.ts:129]: Good catch, I thought I was already skipping STX if it was `bridgeApproval`. I added that code back in [c577043](https://github.com/MetaMask/metamask-extension/pull/28460/commits/c57704325b369653ed235ffe2df2f384549bf43c)

**Review by micaelae — APPROVED** [review]

File: app/scripts/lib/transaction/smart-transactions-mocks.ts (hunk 4 above)
Comment (dan437) [reply app/scripts/lib/transaction/smart-transactions-mocks.ts:33]: If we would use them regularly when working on this feature I suppose we could keep them, but if they would be rarely used I would prefer to delete them to keep the codebase cleaner.

**Review by dan437 — APPROVED** [review]
Nice job! I've shared feedback about some UI improvements in DMs and it will be resolved in upcoming PRs before this feature is enabled in production.

**Pull request merged**